fixsmtpio.8
.TH FIXSMTPIO 8 2020-12-11
.SH NAME
fixsmtpio \- filter SMTP I/O and exit status
.SH SYNOPSIS
.B fixsmtpio
.I prog
.SH DESCRIPTION
.B fixsmtpio
is a proxy for changing how an SMTP service behaves without changing its code.
By default, with no configuration, it changes almost no observable behavior.
.P
In a typical configuration,
.B fixsmtpio
enables TLS for
.BR qmail-smtpd ,
modernizes responses from
.BR ofmipd ,
and makes both programs behave as though they were designed to run under
.BR authup(8) .
.B fixsmtpio
accomplishes this by adjusting certain
client requests,
server responses,
and exit codes.
The
.B "EXIT CODES"
section is important enough to be all the way up here:
.SH "EXIT CODES"
.B fixsmtpio
exits 18 when
.I control/fixsmtpio
exists and fails to parse.
.P
Part of
.BR fixsmtpio 's
job is to return distinct exit codes for distinct conditions.
For instance,
.B authup
needs to distinguish
.B checkpassword
exiting 1 (to reject a password)
from
.B ofmipd
or
.B qmail-smtpd
exiting 1 (on any error).
When running under
.BR authup ,
.B fixsmtpio
can and should be configured to exit uniquely for each SMTP condition:
.TP 3
14
when the server greeting code starts with 4 (temporary failure),
.TP 3
15
when the server greeting code starts with 5 (permanent failure),
.TP 3
16
when the server times out,
and
.TP 3
0
when the server receives EOF from the client.
.SH "ENVIRONMENT VARIABLES"
When running under
.B "sslserver -n"
or
.BR "s6-ucspitlsd" ,
.B fixsmtpio
can offer TLS.
Set
.B UCSPITLS
to the empty string to permit peers to negotiate TLS before transferring messages.
If the peer has negotiated TLS,
.B FIXSMTPIOTLS
will be set for use by filter rules.
.P
When running under
.BR authup ,
.B AUTHUP_USER
will be set for use by filter rules
(such as those recommended in
.BR "EXIT CODES" ).
.P
If
.B DISABLETLS
is set, the presence and value of
.B UCSPITLS
will be ignored.
.P
If
.B NOFIXSMTPIO
is set,
.B fixsmtpio
will simply replace itself with
.IR prog .
.P
If
.B FIXSMTPIODEBUG
is set,
.B fixsmtpio
will show its work.
Lines are prefixed like so:
.TP 3
1: Request received from client
.TP 3
2: Request sent to server
.TP 3
3: Response received from server
.TP 3
4: Response sent to client
.TP 3
D: DATA received from client and sent to server as is
.SH "FILTER RULES"
Filter rules follow the format:
.P
.I [env]:event:[request-prepend]:response-line-glob:[exitcode]:[response]
.P
(Values other than
.I response
may not contain ":", the field separator.)
.TP 5
.I env
.br
(optional)
Environment variable which must be present for the rule to apply.
If empty, none is required.
.TP 5
.I event
.br
(required)
SMTP verb (or "clienteof", "greeting", or "timeout") to which the rule applies.
This field is matched case-insensitively.
.TP 5
.I request-prepend
.br
(optional)
String to prepend to the request before passing it to the server.
"NOOP " (with trailing space) causes the server to trivially accept a request,
performing no action.
If empty, the request is sent as is.
.TP 5
.I response-line-glob
.br
(required)
.BR fnmatch(3) -style
glob pattern, with no special options set.
Each line of the response is compared against this glob.
.TP 5
.I exitcode
.br
(optional)
Numeric code with which
.B fixsmtpio
is to exit immediately.
If empty, it will exit whenever it normally would.
If there are no rules specifying a value for this field,
.B fixsmtpio
will always exit with the same code as its child
.IR prog .
.TP 5
.I response
.br
(optional)
String that replaces every matching line of the server response.

If empty, matching lines are removed.

Strings starting with
"&fixsmtpio"
are reserved for special handling.

"&fixsmtpio_noop"
causes the response to be sent as is.

"&fixsmtpio_fixup"
computes the replacement string via an event-specific internal routine.
(If no corresponding routine exists,
.B fixsmtpio
will refuse to run.)
Routines exist for
.RS 10
.TP 6
.BR HELP :
.br
Prepend this program's home page,
in an attempt to direct support requests to
.BR AUTHOR .
Please enable it!
.TP 6
.BR EHLO ,
.BR HELO ,
.BR QUIT ,
and "greeting":
.br
Include
.I smtpgreeting
in the response,
matching what
.B qmail-smtpd
already does.
This brings
.B ofmipd
up to par.
.RE
.TP 0
.P
Rules are applied in the order written. For instance, if two rules
match, the later rule re-modifies the response returned by the
earlier rule.
If multiple matching rules for an event set
.BR exitcode ,
the last one wins.
.P
Not all rules (alone or in combination) make practical sense.
An earlier rule can cause a later one to start or stop matching.
Keep your configuration as simple as possible, and test it well.
.SH "CONTROL FILES"
.TP 5
.I fixsmtpio
Filter rules as described in
.BR "FILTER RULES" .
Default: none.
.TP 5
.I smtpgreeting
SMTP greeting message.
Default:
.IR me ,
if that is supplied;
otherwise
.B fixsmtpio
will refuse to run.
The first word of
.I smtpgreeting
should be the current host's name.
.SH "COMPATIBILITY"
.B fixsmtpio
must terminate TLS in order to observe and modify requests and responses.
If you rely on
.B qmail-smtpd
being patched to support STARTTLS directly,
please share your use case with the author.
.SH "EXAMPLES"
See
.IR https://schmonz.com/software/acceptutils .
.SH "AUTHOR"
.B Amitai Schleier <schmonz-web-acceptutils@schmonz.com>
.SH "SEE ALSO"
authup(8),
sslserver(1),
s6-ucspitlsd(8),
ucspi-tls(2),
checkpassword(8),
checknotroot(8),
qmail-smtpd(8),
ofmipd(8),
qmail-qfilter-queue(8),
fnmatch(3),
fixcrio(1),
spamdyke.


fixsmtpio.pl
#!/usr/bin/env perl

use warnings;
use strict;

use POSIX qw(_exit);

sub die_usage { die "usage: fixsmtpio.pl prog [ arg ... ]\n"; }
sub die_pipe  { die "fixsmtpio.pl: unable to open pipe: $!\n"; }
sub die_fork  { die "fixsmtpio.pl: unable to fork: $!\n"; }
sub die_read  { die "fixsmtpio.pl: unable to read: $!\n"; }
sub die_write { die "fixsmtpio.pl: unable to write: $!\n"; }

# XXX struct request_response

my $exitcode = 0;

sub strip_last_eol { my ($string_ref) = @_;
    ${$string_ref} =~ s/\r?\n$//;
}

sub accepted_data { my ($response) = @_;
    return $response =~ /^354 /;
}

sub munge_timeout { my ($response_ref) = @_;
    $exitcode = 16;
}

sub munge_greeting { my ($response_ref) = @_;
    if (${$response_ref} =~ /^4/) {
        $exitcode = 14;
    } elsif (${$response_ref} =~ /^5/) {
        ${$response_ref} = 15;
    } else {
        ${$response_ref} = q{235 ok};
        ${$response_ref} .= qq{, $ENV{AUTHUSER},} if defined $ENV{AUTHUSER};
        ${$response_ref} .= qq{ go ahead $< (#2.0.0)};
    }
}

sub munge_help { my ($response_ref) = @_;
    ${$response_ref} = q{214 fixsmtpio.pl home page: }
        . q{https://schmonz.com/qmail/authutils}
        . "\r\n"
        . ${$response_ref};
}

sub munge_test { my ($response_ref) = @_;
    ${$response_ref} .= q{ and also it's mungeable};
}

sub munge_ehlo { my ($response_ref) = @_;
    # XXX @avoids
    my @lines = split(/\r\n/, ${$response_ref});
    @lines = grep { ! /^250.AUTH / } @lines;
    @lines = grep { ! /^250.STARTTLS/ } @lines;
    ${$response_ref} = join("\r\n", @lines);
}

sub verb_matches { my ($s, $sa) = @_;
    return 0 unless length $sa;
    return lc $s eq lc $sa;
}

sub change_every_line_fourth_char_to_dash { my ($multiline_ref) = @_;
    ${$multiline_ref} =~ s/^(.{3})./$1-/msg;
}

sub change_last_line_fourth_char_to_space { my ($multiline_ref) = @_;
    ${$multiline_ref} =~ s/(.{3})-(.+)$/$1 $2/;
}

sub reformat_multiline_response { my ($response_ref) = @_;
    # XXX maybe fix up line breaks? or just keep them correct all along
    change_every_line_fourth_char_to_dash($response_ref);
    change_last_line_fourth_char_to_space($response_ref);
    ${$response_ref} .= "\r\n";
}

sub munge_response { my ($response_ref, $verb, $arg) = @_;
    strip_last_eol($response_ref);
    munge_timeout($response_ref) if '' eq $verb;
    munge_greeting($response_ref) if verb_matches('greeting', $verb);
    munge_help($response_ref) if verb_matches('help', $verb);
    munge_test($response_ref) if verb_matches('test', $verb);
    munge_ehlo($response_ref) if verb_matches('ehlo', $verb);
    reformat_multiline_response($response_ref);
}

sub is_entire_line { my ($string) = @_;
    return substr($string, -1, 1) eq "\n";
}

sub could_be_final_response_line { my ($line) = @_;
    return length($line) >= 4 && substr($line, 3, 1) eq " ";
}

sub is_entire_response { my ($response) = @_;
    my @lines = split(/\r\n/, $response);
    return could_be_final_response_line($lines[-1]) && is_entire_line($response);
}

sub use_as_stdin { my ($fd) = @_;
    open(STDIN, '<&=', $fd) || die_pipe();
}

sub use_as_stdout { my ($fd) = @_;
    open(STDOUT, '>>&=', $fd) || die_pipe();
}

sub mypipe { my ($from_ref, $to_ref) = @_;
    pipe(${$from_ref}, ${$to_ref}) or die_pipe();
}

sub setup_server { my ($from_proxy, $to_server,
                       $from_server, $to_proxy) = @_;
    close($from_server);
    close($to_server);
    use_as_stdin($from_proxy);
    use_as_stdout($to_proxy);
}

sub exec_server_and_never_return { my (@argv) = @_;
    exec @argv;
    die;
}

sub be_child { my ($from_proxy, $to_proxy,
                   $from_server, $to_server,
                   @args) = @_;
    setup_server($from_proxy, $to_server, $from_server, $to_proxy);
    exec_server_and_never_return(@args);
}

sub setup_proxy { my ($from_proxy, $to_proxy) = @_;
    close($from_proxy);
    close($to_proxy);
}

my $want_to_read_bits;

sub want_to_read { my ($fd) = @_;
    vec($want_to_read_bits, fileno($fd), 1) = 1;
}

my $can_read_bits;

sub can_read { my ($fd) = @_;
    return 1 == vec($can_read_bits, fileno($fd), 1);
}

sub can_read_something {
    my $ready;
    $ready = select($can_read_bits = $want_to_read_bits, undef, undef, undef);
    die_read() if $ready == -1 && ! $!{EINTR};
    return $ready;
}

sub saferead { my ($fd, $buf_ref, $len) = @_;
    my $r;
    $r = sysread($fd, ${$buf_ref}, $len);
    die_read() if $r == -1 && ! $!{EINTR};
    return $r;
}

sub safeappend { my ($string_ref, $fd) = @_;
    my ($r, $buf);
    $r = saferead($fd, \$buf, 128);
    ${$string_ref} .= $buf;
    return $r;
}

sub is_last_line_of_data { my ($r) = @_;
    return $r =~ /^\.\r$/;
}

sub parse_request { my ($request_ref, $verb_ref, $arg_ref) = @_;
    strip_last_eol($request_ref);

    (${$verb_ref}, ${$arg_ref}) = split(/ /, ${$request_ref}, 2);
    ${$verb_ref} ||= '';
    ${$arg_ref}  ||= '';

    ${$request_ref} .= "\r\n";
}

sub logit { my ($logprefix, $s) = @_;
    print STDERR "$logprefix: $s";
}

sub write_to_client { my ($fd, $s) = @_;
    syswrite($fd, $s) || die_write();
    logit('O', $s);
}

sub write_to_server { my ($fd, $s) = @_;
    syswrite($fd, $s) || die_write();
    logit('I', $s);
}

sub smtp_test { my ($verb, $arg) = @_;
    return "250 fixsmtpio.pl test ok: $arg";
}

sub smtp_unimplemented { my ($verb, $arg) = @_;
    return "502 unimplemented (#5.5.1)";
}

sub handle_internally { my ($request, $verb_ref, $arg_ref) = @_;
    return "" if verb_matches('noop', ${$verb_ref});
    return smtp_test(${$verb_ref}, ${$arg_ref}) if verb_matches('test', ${$verb_ref});
    return smtp_unimplemented(${$verb_ref}, ${$arg_ref}) if verb_matches('auth', ${$verb_ref});
    return smtp_unimplemented(${$verb_ref}, ${$arg_ref}) if verb_matches('starttls', ${$verb_ref});

    return "";
}

sub send_keepalive { my ($server, $request) = @_;
    write_to_server($server, "NOOP fixsmtpio.pl " . $request);
}

sub check_keepalive { my ($client, $response) = @_;
    if ($response !~ /^250 ok/) {
        write_to_client($client, $response);
        die();
    }
}

sub blocking_line_read { my ($fd) = @_;
    my $line = <$fd>;
    return $line;
}

sub handle_request { my ($from_client, $to_server,
                         $from_server, $to_client,
                         $request, $verb_ref, $arg_ref,
                         $want_data_ref, $in_data_ref) = @_;
    my $internal_response;
    my $keepalive_response;

    if (${$in_data_ref}) {
        write_to_server($to_server, $request);
        ${$in_data_ref} = 0 if is_last_line_of_data($request);
    } else {
        if ($internal_response = handle_internally($request, $verb_ref, $arg_ref)) {
            send_keepalive($to_server, $request);
            $keepalive_response = blocking_line_read($from_server);
            check_keepalive($to_client, $keepalive_response);
            logit('O', $keepalive_response);

            logit('I', $request);
            munge_response(\$internal_response, ${$verb_ref}, ${$arg_ref});
            write_to_client($to_client, $internal_response);
            (${$verb_ref}, ${$arg_ref}) = ('', '');
        } else {
            ${$want_data_ref} = 1 if (verb_matches('data', ${$verb_ref}));
            write_to_server($to_server, $request);
        }
    }
}

sub handle_response { my ($to_client,
                          $response, $verb, $arg,
                          $want_data_ref, $in_data_ref) = @_;
    if (${$want_data_ref}) {
        ${$want_data_ref} = 0;
        ${$in_data_ref} = 1 if accepted_data($response);
    }
    munge_response(\$response, $verb, $arg);
    write_to_client($to_client, $response);
}

# XXX sub request_response_init()

sub do_proxy_stuff { my ($from_client, $to_server,
                         $from_server, $to_client) = @_;
    my ($partial_request, $partial_response) = ('', '');
    my ($request, $verb, $arg, $response) = ('', '', '', '');
    my ($want_data, $in_data) = (0, 0);

    $request = 'greeting';

    want_to_read($from_client);
    want_to_read($from_server);
    for (;;) {
        if (length $request && ! length $response) {
            parse_request(\$request, \$verb, \$arg);
            handle_request($from_client, $to_server,
                           $from_server, $to_client,
                           $request, \$verb, \$arg,
                           \$want_data, \$in_data);
        }

        if (length $response) {
            handle_response($to_client,
                            $response, $verb, $arg,
                            \$want_data, \$in_data);
            ($request, $verb, $arg, $response) = ('', '', '', '');
        }

        next unless can_read_something();

        if (can_read($from_client)) {
            last unless safeappend(\$partial_request, $from_client);
            if (is_entire_line($partial_request)) {
                $request = $partial_request;
                $partial_request = '';
            }
        }

        if (can_read($from_server)) {
            last unless safeappend(\$partial_response, $from_server);
            if (is_entire_response($partial_response)) {
                $response = $partial_response;
                $partial_response = '';
            }
        }
    }
}

sub wait_crashed { my ($wstat) = @_;
    # XXX untested
    return $wstat & 127;
}

sub teardown_proxy_and_exit { my ($child, $from_server, $to_server) = @_;
    my $wstat;

    close($from_server);
    close($to_server);

    die if (waitpid($child, 0) == -1); $wstat = $?;
    die if (wait_crashed($wstat));

    _exit($exitcode);
}

sub be_parent { my ($from_client, $to_client,
                    $from_proxy, $to_proxy,
                    $from_server, $to_server,
                    $child) = @_;
    setup_proxy($from_proxy, $to_proxy);
    do_proxy_stuff($from_client, $to_server, $from_server, $to_client);

    teardown_proxy_and_exit($child, $from_server, $to_server);
}

sub main { my (@args) = @_;
    my $from_client;
    my ($from_proxy, $to_server);
    my ($from_server, $to_proxy);
    my $to_client;
    my $child;

    die_usage() unless @args >= 1;

    $from_client = \*STDIN;
    mypipe(\$from_proxy, \$to_server);
    mypipe(\$from_server, \$to_proxy);
    $to_client = \*STDOUT;

    if ($child = fork()) {
        be_parent($from_client, $to_client,
                  $from_proxy, $to_proxy,
                  $from_server, $to_server,
                  $child);
    } elsif (defined $child) {
        be_child($from_proxy, $to_proxy,
                 $from_server, $to_server,
                 @args);
    } else {
        die_fork();
    }
}

main(@ARGV);
